README.md
# Ambient Extractor

Fork of [color_extractor](https://www.home-assistant.io/integrations/color_extractor/), adding automatic brightness.

Like color_extractor, this integration will extract the predominant color from a given image and apply it to a target light. 
Additionally, overall brightness can be calculated and applied within adjustable limits. Useful as part of an automation.

### Service data attributes
| Attribute | Optional | Type | Default | Description |
|--|--|--|--|--|
| ambient_extract_url | * | URI | - | The full URL (including schema, `http://`, `https://`) of the image to process
| ambient_extract_path | * | String | - | The full path to the image file on local storage we’ll process
| entity_id | No | String | - | The light(s) we’ll set color and/or brightness of
| color_temperature | Yes | Int: 1.000 to 40.000 | 6.600 K | Apply color temperature correction
| brightness_auto | Yes | Boolean | False | Detect and set brightness
| brightness_mode | Yes | mean rms natural dominant | mean | Brightness calculation method. `mean` and `rms` use a grayscale image, `natural` uses perceived brightness, `dominant` the same color as for RGB (fastest).
| brightness_min  | Yes | Int: 0 to 255 | 2 | Minimal brightness. `< 2` means off for most devices.
| brightness_max  | Yes | Int: 0 to 255 | 70 | Maximal brightness, should be `> brightness_min`.
| crop_left | Yes | Int: 0 to 99 | 0 | Crop area: Left offset in % of image width. Default: 0
| crop_top | Yes | Int: 0 to 99 | 0 | Crop area: Top offset in % of image height. Default: 0
| crop_width | Yes | Int: 0 to 100 | 0 | Crop area: Width. Default: 0 (= no cropping)
| crop_height | Yes | Int: 0 to 100 | 0 | Crop area: Height. Default: 0 (= no cropping)

*) Either `ambient_extract_url`or `ambient_extract_path`needs to be set. 

**Please ensure any [external URLs](https://www.home-assistant.io/docs/configuration/basic/#allowlist_external_urls) or [external files](https://www.home-assistant.io/docs/configuration/basic/#allowlist_external_dirs) are authorized for use, you will receive error messages if this component is not allowed access to these external resources.**

Besides `rgb_color`and `brightness`, feel free to set [generic light](https://www.home-assistant.io/integrations/light/) attributes. For a static brightness setting, don't enable `brightness_auto`, just add a `brightness: ` or `brightness_pct:` value.

### Automation trigger recommendations

Slow sources like Android Debug Bridge (ADB) can take up to 15 seconds for a fully sized screenshot.
```yaml
trigger:
  - platform: time_pattern
    seconds: "*/15"
    minutes: "*"
    hours: "*"
  - platform: state
    entity_id: media_player.firetv
```

Ideal conditions using a fast source and scaled down images may allow for 2-3 times per second.
Enigma2 example: `http://enigma2/grab?format=png&mode=video&r=64`.
When using multiple ZHA light entities, consider creating a ZHA group to off-load your ZigBee network. 


## Installation

### Using HACS

1. Ensure that [HACS](https://github.com/hacs/integration) is installed.
2. Add Custom repository `https://github.com/xplus2/homeassistant-ambient-extractor`. Category `Integration`.
3. Install the "Ambient Extractor" integration.
4. Restart Home Assistant.

### Manual installation

1. Copy the folder `ambient_extractor` to `custom_components` in your Home Assistant `config` folder.
2. [Configure the integration](#configuration).
3. Restart Home Assistant.

## Configuration
Add the following line to your `configuration.yaml` (not needed when using HACS):

    ambient_extractor:


## Usage examples

```yaml
service: ambient_extractor.turn_on
data_template:
  ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=96"
  entity_id:
    - light.living_room_zha_group_0x0002
  transition: 0.4
  
  # bool, default: false
  brightness_auto: true
  
  # string(mean|rms|natural), default: mean
  brightness_mode: natural
  
  # 0-255, default: 2
  brightness_min: 2
  
  # 0-255, default: 70
  brightness_max: 70
```

### Using helper variables

```yaml
service: ambient_extractor.turn_on
data_template:
  ambient_extract_url: "{{ states.media_player.firetv.attributes.entity_picture }}"
  entity_id:
    - light.living_room_zha_group_0x0002
  transition: 0.3
  color_temperature: "{{ states('input_number.ambilight_color_temperature') }}"
  brightness_auto: true
  brightness_mode: natural
  brightness_min: "{{ states('input_number.ambilight_brightness_min') }}"
  brightness_max: "{{ states('input_number.ambilight_brightness_max') }}"
```
Create `ambilight_color_temperature` as Number from 1.000 to 40.000, step size 1.


### Full automation YAML


#### Using a fast image source

Two times per second, if screenshots can be accessed fast enough. Tested with OpenATV on Vu+ Uno 4K SE.

```yaml
alias: Ambient Light enigma2
description: ""
trigger:
  - platform: time_pattern
    seconds: "*"
    minutes: "*"
    hours: "*"
condition:
  - condition: state
    entity_id: media_player.enigma2
    state: playing
action:
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=96"
      entity_id:
        - light.living_room_zha_group_0x0001
      transition: 0.3
      brightness_auto: true
  - delay:
      hours: 0
      minutes: 0
      seconds: 0
      milliseconds: 350
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=96"
      entity_id:
        - light.living_room_zha_group_0x0001
      transition: 0.3
      brightness_auto: true
mode: single
```

Left, right and ceiling using crop_*.

```yaml
alias: Ambient Light enigma2
description: ""
trigger:
  - platform: time_pattern
    seconds: "*"
    minutes: "*"
    hours: "*"
condition:
  - condition: state
    entity_id: media_player.enigma2
    state: playing
action:
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=64"
      entity_id:
        - light.living_room_tv_left
      transition: 0.6
      brightness_auto: true
      crop_left: 0
      crop_top: 0
      crop_width: 25
      crop_height: 100
  - delay:
      hours: 0
      minutes: 0
      seconds: 0
      milliseconds: 200
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=64"
      entity_id:
        - light.living_room_tv_right
      transition: 0.6
      brightness_auto: true
      crop_left: 75
      crop_width: 25
      crop_height: 100
  - delay:
      hours: 0
      minutes: 0
      seconds: 0
      milliseconds: 200
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "http://enigma2/grab?format=png&mode=video&r=64"
      entity_id:
        - light.living_room_ceiling
      transition: 0.6
      brightness_auto: true
      crop_width: 100
      crop_height: 35
mode: single
```

`crop_width` and `crop_height` need values > 0 or cropping is disabled.

#### Using slower sources
```yaml
alias: Ambient Light FireTV
description: ""
trigger:
  - platform: state
    entity_id: media_player.firetv
action:
  - service: ambient_extractor.turn_on
    data_template:
      ambient_extract_url: "{{ states.media_player.firetv.attributes.entity_picture }}"
      entity_id:
        - light.living_room_floor_lamp
      transition: 2
      brightness_auto: true
      color_temperature: 5000
mode: single
```


CHANGELOG.md
# v0.6.1
`color_temperature` now accepts values from 1.000 to 40.000 K, step size 1.
New default is 6.600

# v0.6
New features:
- color_temperature
  When using a receiver or other player as source, apply the tv's color temperature.

  Configuration:
  - color_temperature = []
# v0.5.2
Expanded crop functionality to color detection.
Changed "crop_offset_left" to "offset_left".
Changed "crop_offset_top" to "offset_top".

# v0.5.1
Fixed services.yaml

# v0.5
New features: 
- Limit brightness detection to an area of the source image.

  Configuration:
  - crop_offset_left = [0-100] in % width; default = 0
  - crop_offset_top = [0-100] in % height; default = 0
  - crop_width = [0-100] in % width; default = 0 == full image
  - crop_height = [ß-100] in % height; default = 0 == full image

# v0.4

Released 2023-02-20

- Removed experimental rgb_temperature

# v0.3

Released 2022-12-13

- Added rgb_temperature (-250; 250). Apply color correction on RGB values.

# v0.2

Released 2022-12-13

- Changed URL

# v0.1

Released 2022-12-13

- Initial release



custom_components/ambient_extractor/extract_color.py
from tempfile import TemporaryFile
from colorthief import ColorThief
from PIL import Image
import math


def get_file(file_path):
    """Get a PIL acceptable input file reference.

    Allows us to mock patch during testing to make BytesIO stream.
    """
    return file_path


def get_cropped_image(file_handler, crop_area):
    im = Image.open(file_handler)
    if crop_area['active']:
        im_width, im_height = im.size
        im = im.crop((
            math.floor(im_width / 100 * crop_area['x']),
            math.floor(im_height / 100 * crop_area['y']),
            math.floor(im_width / 100 * (crop_area['x'] + crop_area['w'])),
            math.floor(im_width / 100 * (crop_area['y'] + crop_area['h'])),
        ))
    return im


def get_color_from_image(im) -> tuple:
    file_handler = TemporaryFile()
    im.save(file_handler, "PNG")
    return get_color_from_file(file_handler)


def get_color_from_file(file_handler) -> tuple:
    """Given an image file, extract the predominant color from it."""
    color_thief = ColorThief(file_handler)

    # get_color returns a SINGLE RGB value for the given image
    color = color_thief.get_color(quality=1)
    return color




custom_components/ambient_extractor/color_temperature.py
import math


# @see https://tannerhelland.com/2012/09/18/convert-temperature-rgb-algorithm-code.html
def apply_color_temperature(color, color_temperature) -> tuple:
    if color_temperature == 6600:
        return color

    factor_red = 255 if color_temperature <= 6600 \
        else 329.698727446 * ((color_temperature / 100 - 60) ** -0.1332047592)
    factor_green = 99.4708025861 * math.log(color_temperature / 100) - 161.1195681661 if color_temperature <= 6600 \
        else 288.1221695283 * ((color_temperature / 100 - 60) ** -0.0755148492)
    factor_blue = 255 if color_temperature >= 6600 else 0 if color_temperature <= 1900 \
        else 138.5177312231 * math.log(color_temperature / 100 - 10) - 305.0447927307

    factor_red = 0 if factor_red < 0 else 255 if factor_red > 255 else factor_red
    factor_green = 0 if factor_green < 0 else 255 if factor_green > 255 else factor_green
    factor_blue = 0 if factor_blue < 0 else 255 if factor_blue > 255 else factor_blue

    r, g, b = color
    return (
        r * (factor_red / 255.0),
        g * (factor_green / 255.0),
        b * (factor_blue / 255.0)
    )


custom_components/ambient_extractor/__init__.py
"""Module for ambient_extractor (RGB/brightness extraction from images) component."""
import asyncio
import io
import logging
from PIL import UnidentifiedImageError
import aiohttp
import async_timeout
import voluptuous as vol

from homeassistant.components.light import (
    ATTR_RGB_COLOR,
    ATTR_BRIGHTNESS,
    DOMAIN as LIGHT_DOMAIN,
    LIGHT_TURN_ON_SCHEMA,
)
from homeassistant.const import SERVICE_TURN_ON as LIGHT_SERVICE_TURN_ON
from homeassistant.core import HomeAssistant, ServiceCall
from homeassistant.helpers import aiohttp_client
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.typing import ConfigType

from .const import (
    DOMAIN,
    SERVICE_TURN_ON,
    ATTR_PATH,
    ATTR_URL,
    ATTR_COLOR_TEMPERATURE,
    ATTR_BRIGHTNESS_AUTO,
    ATTR_BRIGHTNESS_MODE,
    ATTR_BRIGHTNESS_MIN,
    ATTR_BRIGHTNESS_MAX,
    ATTR_CROP_X,
    ATTR_CROP_Y,
    ATTR_CROP_W,
    ATTR_CROP_H,
)
from .color_temperature import apply_color_temperature
from .extract_brightness import get_brightness
from .extract_color import get_file, get_color_from_image, get_color_from_file, get_cropped_image

_LOGGER = logging.getLogger(__name__)

# Extend the existing light.turn_on service schema
SERVICE_SCHEMA = vol.All(
    cv.has_at_least_one_key(ATTR_URL, ATTR_PATH),
    cv.make_entity_service_schema({
        **LIGHT_TURN_ON_SCHEMA,
        vol.Exclusive(ATTR_PATH, "ambient_extractor"): cv.isfile,
        vol.Exclusive(ATTR_URL, "ambient_extractor"): cv.url,
        vol.Optional(ATTR_COLOR_TEMPERATURE, default=False): cv.positive_int,
        vol.Optional(ATTR_BRIGHTNESS_AUTO, default=False): cv.boolean,
        vol.Optional(ATTR_BRIGHTNESS_MODE, default="mean"): cv.string,
        vol.Optional(ATTR_BRIGHTNESS_MIN, default=2): cv.positive_int,
        vol.Optional(ATTR_BRIGHTNESS_MAX, default=70): cv.positive_int,
        vol.Optional(ATTR_CROP_X, default=0): cv.positive_int,
        vol.Optional(ATTR_CROP_Y, default=0): cv.positive_int,
        vol.Optional(ATTR_CROP_W, default=0): cv.positive_int,
        vol.Optional(ATTR_CROP_H, default=0): cv.positive_int,
    }),
)


async def async_setup(hass: HomeAssistant, hass_config: ConfigType) -> bool:

    async def async_handle_service(service_call: ServiceCall) -> None:
        service_data = dict(service_call.data)

        br_min = 2
        br_max = 70
        check_brightness = True
        br_mode = "mean"
        if ATTR_BRIGHTNESS_MIN in service_data:
            br_min = service_data.pop(ATTR_BRIGHTNESS_MIN)
        if ATTR_BRIGHTNESS_MAX in service_data:
            br_max = service_data.pop(ATTR_BRIGHTNESS_MAX)
        if ATTR_BRIGHTNESS_AUTO in service_data:
            check_brightness = service_data.pop(ATTR_BRIGHTNESS_AUTO)
        if ATTR_BRIGHTNESS_MODE in service_data:
            br_mode = service_data.pop(ATTR_BRIGHTNESS_MODE)

        crop_area = {
            'active': False,
            'x': 0,
            'y': 0,
            'w': 0,
            'h': 0,
        }
        if ATTR_CROP_X in service_data:
            crop_area['x'] = service_data.pop(ATTR_CROP_X)
        if ATTR_CROP_Y in service_data:
            crop_area['y'] = service_data.pop(ATTR_CROP_Y)
        if ATTR_CROP_W in service_data:
            crop_area['w'] = service_data.pop(ATTR_CROP_W)
        if ATTR_CROP_H in service_data:
            crop_area['h'] = service_data.pop(ATTR_CROP_H)

        color_temperature = 6600
        if ATTR_COLOR_TEMPERATURE in service_data:
            color_temperature = service_data.pop(ATTR_COLOR_TEMPERATURE)
            color_temperature = 1000 if color_temperature < 1000 \
                else 40000 if color_temperature > 40000 \
                else color_temperature

        # Don't crop if height or width == 0
        if crop_area['w'] > 0 and crop_area['h'] > 0:
            crop_area['active'] = True
            if crop_area['x'] + crop_area['w'] > 100:
                crop_area['w'] = 100 - crop_area['x']
            if crop_area['y'] + crop_area['h'] > 100:
                crop_area['h'] = 100 - crop_area['y']

        try:
            if ATTR_URL in service_data:
                image_type = "URL"
                image_reference = service_data.pop(ATTR_URL)
                colorset = await async_extract_color_from_url(
                    image_reference, check_brightness, br_mode, crop_area
                )

            elif ATTR_PATH in service_data:
                image_type = "file path"
                image_reference = service_data.pop(ATTR_PATH)
                colorset = await hass.async_add_executor_job(
                    extract_color_from_path, image_reference, check_brightness, br_mode, crop_area
                )

            color = colorset["color"]
            if check_brightness:
                brightness = colorset["brightness"]

        except UnidentifiedImageError as ex:
            _LOGGER.error("Bad image from %s '%s' provided. %s", image_type, image_reference, ex)
            return

        if color:
            color = apply_color_temperature(color, color_temperature)
            service_data[ATTR_RGB_COLOR] = color

        if brightness:
            service_data[ATTR_BRIGHTNESS] = br_min if br_min >= br_max \
                else br_min + ((brightness / 255) * (br_max - br_min))

        if color or brightness:
            await hass.services.async_call(
                LIGHT_DOMAIN, LIGHT_SERVICE_TURN_ON, service_data, blocking=True
            )

    hass.services.async_register(
        DOMAIN,
        SERVICE_TURN_ON,
        async_handle_service,
        schema=SERVICE_SCHEMA,
    )

    async def async_extract_color_from_url(url, check_brightness, br_mode, crop_area):
        """Handle call for URL based image."""
        if not hass.config.is_allowed_external_url(url):
            _LOGGER.error("External URL '%s' is not allowed, please add to 'allowlist_external_urls'", url)
            return None

        try:
            session = aiohttp_client.async_get_clientsession(hass)
            async with async_timeout.timeout(10):
                response = await session.get(url)

        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
            _LOGGER.error("Failed to get image due to HTTPError: %s", err)
            return None

        content = await response.content.read()
        with io.BytesIO(content) as _file:
            _file.name = "ambient_extractor.jpg"
            _file.seek(0)
            im = get_cropped_image(_file, crop_area)
            color = get_color_from_image(im) if crop_area['active'] else get_color_from_file(_file)
            brightness = get_brightness(im, br_mode, color) if check_brightness else 0
            return {
                "color": color,
                "brightness": brightness
            }

    def extract_color_from_path(file_path, check_brightness, br_mode, crop_area):
        """Handle call for local file based image."""
        if not hass.config.is_allowed_path(file_path):
            _LOGGER.error("File path '%s' is not allowed, please add to 'allowlist_external_dirs'", file_path)
            return None

        _file = get_file(file_path)
        im = get_cropped_image(_file, crop_area)
        color = get_color_from_image(im) if crop_area['active'] else get_color_from_file(_file)
        brightness = get_brightness(im, br_mode, color) if check_brightness else 0
        return {
            "color": color,
            "brightness": brightness
        }

    return True
